src/theory/quantifiers/sygus_grammar_norm.h

```diff
@@ -14,137 +14,407 @@
   TermDbSygus* d_tds;
+  /** List of variable inputs of function-to-synthesize.
+   *
+   * This information is needed in the construction of each datatype
+   * representation of type nodes contained in the type node being normalized
+   */
+  TNode d_sygus_vars;
+  /* Datatypes to be resolved */
+  std::vector<Datatype> d_dt_all;
+  /* Types to be resolved */
+  std::set<Type> d_unres_t_all;
+  /* Associates type nodes with OpPosTries */
+  std::map<TypeNode, OpPosTrie> d_tries;
+  /* Map of type nodes into their identity operators (\lambda x. x) */
+  static std::map<TypeNode, Node> d_tn_to_id;
 
-  /** normalize integer type
+  /** recursively traverses a typenode normalizing all of its elements
    *
-   * TODO actually perform the normalization #1304
+   * "tn" is the typenode to be normalized
+   * "dt" is its datatype representation
+   * "op_pos" is the list of positions of construtors of dt that are being
+   * considered for the normalization
    *
-   * ind is the index of the analyzed typeobject in tos
+   * The result of normalizing tn with the respective constructors is cached
+   * with an OpPosTrie. New types and datatypes created during normalization are
+   * accumulated grobally to be later resolved.
    *
-   * New types created during normalization will be added to tos and
-   * tn_to_unres
+   * The normalization occurs following some inferred transformation based on
+   * the sygus type (e.g. Int) of tn, and the operators being considered.
    *
-   * sygus_vars is used as above for datatype construction */
-  void normalizeSygusInt(unsigned ind,
-                         std::vector<TypeObject>& tos,
-                         std::map<TypeNode, Type>& tn_to_unres,
-                         Node sygus_vars);
-
-  /** Traverses the datatype representation of src_tn and collects the types it
-   * contains
+   * Example: Let A be the type node encoding the grammar
    *
-   * For each new typenode a TypeObject is created, with an unresolved type and
-   * a datatype to be later resolved and constructed, respectively. These are
-   * accumulated in tos.
+   * Int -> x | y | Int + Int | 0 | 1 | ite(Bool, Int, Int)
    *
-   * tn_to_unres maps the sygus types that the typenodes stand for into
-   * the unresolved type to be built for the typenodes normalizations.
+   * and assume all its datatype constructors are being used for
+   * normalization. The inferred normalization transformation will consider the
+   * non-zero elements {x, y, 1, ite(...)} and the operator {+} to build a chain
+   * of monomials, as seen above. The operator for "0" is rebuilt as is (the
+   * default behaviour of operators not selected for transformations).
+   *
+   * recursion depth is limited by the height of the types, which is small
+   */
+  TypeNode normalizeSygusRec(TypeNode tn,
+                             const Datatype& dt,
+                             std::vector<unsigned> op_pos);
```

Comment:
op_pos should be const std::vector<unsigned>&

Reply:
I can't pass this as ``const`` because ``op_pos`` will be passed to a function that will modify it, but I'm now passing as a reference.